packages/server/src/http/resolveHTTPResponse.ts

```diff
@@ -43,70 +42,151 @@ interface ResolveHTTPRequestOptions<
+    | undefined,
+  errors: TRPCError[] = [],
+): HTTPResponse {
+  let status = untransformedJSON ? getHTTPStatusCode(untransformedJSON) : 200;
+  const headers: HTTPHeaders = {
+    'Content-Type': 'application/json',
+  };
+
+  const meta =
+    responseMeta?.({
+      ctx,
+      paths,
+      type,
+      data: untransformedJSON
+        ? Array.isArray(untransformedJSON)
+          ? untransformedJSON
+          : [untransformedJSON]
+        : untransformedJSON,
+      errors,
+    }) ?? {};
+
+  for (const [key, value] of Object.entries(meta.headers ?? {})) {
+    headers[key] = value;
+  }
+  if (meta.status) {
+    status = meta.status;
+  }
+
+  return {
+    status,
+    headers,
+  };
+}
+
+async function inputToProcedureCall<
   TRouter extends AnyRouter,
   TRequest extends HTTPRequest,
->(opts: ResolveHTTPRequestOptions<TRouter, TRequest>): Promise<HTTPResponse> {
+>(
+  opts: ResolveHTTPRequestOptions<TRouter, TRequest>,
+  ctx: inferRouterContext<TRouter> | undefined,
+  type: Exclude<
+    (typeof HTTP_METHOD_PROCEDURE_TYPE_MAP)[string],
+    undefined | 'subscription'
+  >,
+  inputs: Record<number, unknown>,
+  path: string,
+  index: number,
+): Promise<
+  [index: number, response: TRPCResponse<unknown, inferRouterError<TRouter>>]
+> {
+  const input = inputs[index];
+  try {
+    const data = await callProcedure({
+      procedures: opts.router._def.procedures,
+      path,
+      rawInput: input,
+      ctx,
+      type,
+    });
+    return [
+      index,
+      {
+        result: {
+          data,
+        },
+      },
+    ];
+  } catch (cause) {
+    const error = getTRPCErrorFromUnknown(cause);
+
+    opts.onError?.({ error, path, input, ctx, type: type, req: opts.req });
+
+    return [
+      index,
+      {
+        error: getErrorShape({
+          config: opts.router._def._config,
+          error,
+          type,
+          path,
+          input,
+          ctx,
+        }),
+      },
+    ];
+  }
+}
+
+/**
+ * This generator function has 2 ways of "returning data": what it yields and what it returns.
+ * It also has 2 "steps": the first yield, reserved for header / status (`HTTPResponse`),
+ * and the following yield/return, reserved for the body (`ResponseChunk`).
+ *
+ * - The first yield is always a `HTTPResponse`, and there should always be a first yield.
+ *
+ * After the first yield, you can either:
+ * - return a response chunk (`ResponseChunk`), with index -1 (`[-1, string]`),
+ * signaling that this is the entire response, no streaming.
+ *
+ * or
+ *
+ * - yield as many response chunks as needed (zero chunks is allowed for a response without a body),
+ * and signal the end of the response by returning `undefined`.
+ */
+export async function* resolveHTTPResponse<
+  TRouter extends AnyRouter,
+  TRequest extends HTTPRequest,
+>(
+  opts: ResolveHTTPRequestOptions<TRouter, TRequest>,
+): AsyncGenerator<
+  ResponseChunk | HTTPResponse,
+  ResponseChunk | undefined,
+  never
+> {
```

Comment:
This is a breaking change (we export this from `@trpc/server/http` and community adapters use it.

Reply:
Should I find another way then? Or are we going to make a major semver for this? Or other alternative?

Reply:
BTW, the implementation in adapters/aws-lambda actually just consumes the iterator and returns header + status + body. So this would be a good "retro-compatible" example.

Reply:
Ok to make life easier for everyone, I think I should remove the async generators from the server-side as well. I'll have a PR for this in a few minutes.

Reply:
https://github.com/trpc/trpc/pull/4433